---
title: "GO analysis"
author: "Jason T. Serviss"
date: "09/06/2017"
output:
  html_document:
    code_folding: hide
    highlight: pygments
    theme: readable
---

```{r libraries, message=FALSE}
source('~/Github/diverseScripts/R/scripts/largeDivergingPalettes.R')
library(acidAdaptedRNAseq)
library(GO.db)
library(GOSim)
library(igraph)
library(tidyverse)
library(ggraph)
library(multipanelfigure)
library(grid)
library(gtable)
library(magrittr)
library(printr)
```

```{r, message=FALSE}
#load GO result and subset signignificant terms
data(topGOresult, package = "acidAdaptedRNAseq")
sig <- topGOresult[topGOresult$classicFisher < 0.05, ]
```

Download the GO graph for the significant terms and run the spin-glass community
detection algorithm. Show the number of detected communities and their size.

Note to self: Check if what getGOgraph is downloading is equal to the ancestors 
via GO.db.
```{r, message=FALSE}
#get GO graph
g <- downloadGOgraph(sig$GO.ID)

#run community analysis
tmp <- collapseGraph(g)
cg <- tmp[[1]]
comSummary <- tmp[[2]] %>%
    left_join(select(sig, GO.ID, classicFisher), by=c("GOID" = "GO.ID"))
```

Calculate and show the number of genes included in each community.
```{r, message = FALSE}
#calculate genes per term
uCommunityNames <- unique(comSummary$communityName)
genesPerCom <- lapply(1:length(uCommunityNames), function(x) {
    bool <- comSummary$communityName == uCommunityNames[x]
    idsToGet <- comSummary[bool, ]$GOID
    genesForIdsString <- sig[sig$GO.ID %in% idsToGet, "ID"]
    genesForIdsList <- strsplit(genesForIdsString, ", ")
    length(unlist(genesForIdsList))
})
names(genesPerCom) <- uCommunityNames

namedListToTibble(genesPerCom) %>%
    setNames(c("communityName", "nrCommunityGenes")) %>%
    inner_join(distinct(select(comSummary, communityName, communityTerm))) %>%
    select(communityTerm, nrCommunityGenes, -communityName)
```

Show the similarity between communities.
```{r, fig.align='center', fig.height=17, fig.width=15, eval=TRUE, message=FALSE, warning=FALSE}
cgSim <- similarity(cg, mode = "all", method = "jaccard")
colnames(cgSim) <- unique(
    comSummary[match(V(cg)$name, comSummary$communityID), ]$communityTerm
)
rownames(cgSim) <- unique(
    comSummary[match(V(cg)$name, comSummary$communityID), ]$communityTerm
)
heatmap(1 - cgSim, margins = c(20,20))
```

Set attributes in the collapsed graph.
```{r}
cg <- addAttributes1(cg, comSummary)
```

Add back all nodes from the original graph with the exception of the community
nodes. Add an edge from each of the added nodes to the corresponding community
node. Plot the final result.

```{r, fig.align='center', fig.height=17, fig.width=15, eval=TRUE, message=FALSE, warning=FALSE}
#add vertices and edges from uncollapsed graph
toAdd <- comSummary %>%
    filter(!GOID %in% communityID) %>%
    filter(classicFisher < 0.05) %>%
    select(GOID, communityID) %>%
    setNames(c("from", "to"))

cg <- addBackVertexAndEdges(cg, toAdd)

#add attributes
cg <- addAttributes2(cg, comSummary)

#plot
plotCollapsedGraph(cg, 10, 5)
collapsed <- plotCollapsedGraph(cg)
mylegend <- g_legend(collapsed)
collapsed <- collapsed + theme(legend.position="none")
```


Plot a heatmap showing gene expression per community in the parental and adapted
cell lines.
```{r, fig.align='center', fig.height=17, fig.width=15, eval=TRUE, message=FALSE, warning=FALSE}
#extract all gene IDs per GO term
genePerTerm <- extractGenesFromGO()

#calculate z-score 
zNorm <- calculateZscore()

#merge data, note that GO graph terms not in sig are included in comSummary
heatmap <- comSummary %>%
    filter(!is.na(classicFisher)) %>%
    full_join(genePerTerm, by = "GOID") %>%
    left_join(zNorm, by = "ID") %>%
    gather(condition, zScore, -communityName, 
           -communityID, -communitySize, -communityTerm,
           -GOID, -authority.score, -Term, 
           -classicFisher, -ID
    )

#run clustering per community to get row order
order <- clusterForOrder(heatmap)

#merge order into summary
heatmap <- left_join(heatmap, order, by = c("communityName", "ID"))


#plot
heatmap$plotCondition <- plotCondRename(heatmap)
heatmap <- plotHeatmap(heatmap)
```

Highlight differential gene expression in communities and terms by plotting the expression levels in parental and adapted cell lines.
```{r, fig.align='center', fig.height=8, fig.width=7, eval=TRUE, message=FALSE, warning=FALSE}
#setup rld expression values for merge
rld <- deResultsRld %>%
    as.data.frame() %>%
    rownames_to_column() %>%
    rename(ID = rowname) %>%
    as_tibble() %>%
    gather(condition, expression, -ID)

#setup data for plot
barDat <- comSummary %>%
    filter(!is.na(classicFisher)) %>%
    full_join(genePerTerm, by = "GOID") %>%
    left_join(rld, by = "ID") %>%
    left_join(geneAnnotation, by = "ID")

barDat$plotCondition <- plotCondRename(barDat)

#Specify terms to plot
terms <- c(
    "positive regulation of autophagy",
    "response to osmotic stress",
    "cell death",
    "positive regulation of metabolic process",
    "negative regulation of cell differentiation",
    "regulation of protein ubiquitination",
    "cell cycle",
    "cell proliferation"
)

#plot for rmarkdown
plotGenes(barDat, terms[1])
plotGenes(barDat, terms[2])
plotGenes(barDat, terms[3])
plotGenes(barDat, terms[4])
plotGenes(barDat, terms[5])
plotGenes(barDat, terms[6])
plotGenes(barDat, terms[7])
plotGenes(barDat, terms[8])

#plot for figure
p1 <- plotGenes(barDat, terms[1]) 
mylegend2 <- g_legend(p1)

p1 <- plotGenes(barDat, terms[1]) + theme(legend.position = "none")
p2 <- plotGenes(barDat, terms[2]) + theme(legend.position = "none")
p3 <- plotGenes(barDat, terms[3]) + theme(legend.position = "none")
p4 <- plotGenes(barDat, terms[4]) + theme(legend.position = "none")
p5 <- plotGenes(barDat, terms[5]) + theme(legend.position = "none")
p6 <- plotGenes(barDat, terms[6]) + theme(legend.position = "none")
p7 <- plotGenes(barDat, terms[7]) + theme(legend.position = "none")
p8 <- plotGenes(barDat, terms[8]) + theme(legend.position = "none")

#set up figure panels
figure1 <- multi_panel_figure(
  width = c(200, 100, 100, 100),
  height = 350,
  rows = 5,
  unit = "mm"
 )

#plot to panels
figure1 %<>% fill_panel(mylegend, column = 1:4, row = 5, label = "")
figure1 %<>% fill_panel(collapsed, column = 1, row = 1:4, label = "A")
figure1 %<>% fill_panel(heatmap, column = 2, row = 1:4, label = "B")
figure1 %<>% fill_panel(p1, column = 3, row = 1, label = "C")
figure1 %<>% fill_panel(p2, column = 3, row = 2, label = "D")
figure1 %<>% fill_panel(p3, column = 3, row = 3, label = "E")
figure1 %<>% fill_panel(p4, column = 3, row = 4, label = "F")
figure1 %<>% fill_panel(p5, column = 4, row = 1, label = "G")
figure1 %<>% fill_panel(p6, column = 4, row = 2, label = "H")
figure1 %<>% fill_panel(p7, column = 4, row = 3, label = "I")
figure1 %<>% fill_panel(p8, column = 4, row = 4, label = "J")

#figure1

#save final figure
ggsave(
    plot = figure1, 
    '~/Github/DeMilitoRNAseq-analysis/inst/DeMilitoGO/figure.pdf', 
    width = 550, 
    height = 350, 
    units = "mm", 
    limitsize = FALSE
)
```